### PR TITLE
feat(v2): add ability disable to open page in browser when start command

### PR DESCRIPTION
**Pull request by lex111**
## Motivation

When developing docs site (`docusaurus start` command), it can be very convenient to turn off auto page opening in a browser.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

- Run `yarn run start --disable-open-browser`

### CODE REVIEW

**Comment by lex111** [comment]
> Is the command opening too many windows for you?

Exactly! Every time command open new window, I don’t think it’s a bug, because how can the plugin find out that the link is already open?

**Comment by yangshun** [comment]
We use the same approach as Create React App. You can try CRA for yourself. If you start the server, wait for the window to open, kill it, then start the server again, it will reuse that open window. Try and see.

Since this is likely a bug, we'll need to know your environment in order to repro.

**Comment by lex111** [comment]
Oops, I didn’t know about this, I’ll check CRA.

**Comment by lex111** [comment]
> On Mac OS X, attempts to reuse an existing Chrome tab via AppleScript.

From [docs](https://github.com/facebook/create-react-app/blob/master/packages/react-dev-utils/README.md#openbrowserurl-string-boolean)

Gotcha! We use Linux, so this does not work for us! Therefore, I ask that you please accept this PR.

**Comment by yangshun** [comment]
I see. I'm agreeable to this but I'd also like Endi's opinions on this PR since he's a Linux user.

**Comment by lex111** [comment]
I proceeded from the assumption that without specifying any flags (as of now) there will be an auto-opening in the browser, and only by providing this flag can this standard behavior be disabled.

If the `open-browser` flag is always `true`, then how can I disable it as a user? It turns out we need to change command in package.json, i.e.

```diff
- "start": "docusaurus start",
+ "start": "docusaurus start --open-browser",
```

Right? I would like this change to affect others less, otherwise this feature will stop working for everyone else. I mean, they need to specify this `open-browser` flag manually.

File: packages/docusaurus/bin/docusaurus.js [hunk 1]
```diff
@@ -70,16 +70,13 @@ cli
...
-  )
-  .action((siteDir = '.', {port, host, hotOnly, disableOpenBrowser}) => {
+  .option('--no-open', 'Do not open page in the browser (default: false)')
+  .action((siteDir = '.', {port, host, hotOnly, noOpen}) => {
     wrapCommand(start)(path.resolve(siteDir), {
       port,
       host,
       hotOnly,
-      disableOpenBrowser,
+      noOpen,
```
Comment (endiliey) [inline packages/docusaurus/bin/docusaurus.js]: the variable name will be named “open” iirc

See https://www.npmjs.com/package/commander#other-option-types-negatable-boolean-and-flagvalue

File: packages/docusaurus-types/src/index.d.ts [hunk 2]
```diff
@@ -49,6 +49,7 @@ export interface StartCLIOptions {
   port: string;
   host: string;
   hotOnly: boolean;
+  noOpen: boolean;
```
Comment (endiliey) [inline packages/docusaurus-types/src/index.d.ts]:
```suggestion
  open: boolean;
```

File: packages/docusaurus/bin/docusaurus.js [hunk 3]
```diff
@@ -70,11 +70,13 @@ cli
...
     'Do not fallback to page refresh if hot reload fails (default: false)',
   )
-  .action((siteDir = '.', {port, host, hotOnly}) => {
+  .option('--no-open', 'Do not open page in the browser (default: false)')
+  .action((siteDir = '.', {port, host, hotOnly, noOpen}) => {
     wrapCommand(start)(path.resolve(siteDir), {
       port,
       host,
       hotOnly,
+      noOpen,
```
Comment (endiliey) [inline packages/docusaurus/bin/docusaurus.js]:
```suggestion
      open,
```

File: packages/docusaurus/src/commands/start.ts [hunk 4]
```diff
@@ -144,7 +144,7 @@ export async function start(
     if (err) {
       console.log(err);
     }
-    openBrowser(openUrl);
+    !cliOptions.noOpen && openBrowser(openUrl);
```
Comment (endiliey) [inline packages/docusaurus/src/commands/start.ts]:
```suggestion
    cliOptions.open && openBrowser(openUrl);
```

**Comment by lex111** [comment]
@endiliey I can’t understand, do you suggest by default **not** to open the page automatically? And do this only by  providing `open` flag?

**Comment by endiliey** [comment]
so if ur option is —no-foo

the variable name will be foo, and its true by default.

iirc right now your noOpen will always be undefined

**Comment by lex111** [comment]
@endiliey ah, now I see, thank you!

**Pull request merged**